src/coreclr/jit/valuenum.cpp

```diff
@@ -10221,6 +10350,50 @@ static bool GetStaticFieldSeqAndAddress(ValueNumStore* vnStore, GenTree* tree, s
     return false;
 }
 
+//----------------------------------------------------------------------------------
+// GetObjectHandleAndOffset: Try to obtain a constant object handle with an offset from
+//    the given tree.
+//
+// Arguments:
+//    vnStore    - ValueNumStore object
+//    tree       - tree node to inspect
+//    byteOffset - [Out] resulting byte offset
+//    pObj       - [Out] constant object handle
+//
+// Return Value:
+//    true if the given tree is a ObjHandle + CNS
+//
+static bool GetObjectHandleAndOffset(ValueNumStore*         vnStore,
+                                     GenTree*               tree,
+                                     ssize_t*               byteOffset,
+                                     CORINFO_OBJECT_HANDLE* pObj)
+{
+
+    if (!tree->gtVNPair.BothEqual())
+    {
+        return false;
+    }
+
+    ValueNum  treeVN = tree->gtVNPair.GetLiberal();
+    VNFuncApp funcApp;
+    if (vnStore->GetVNFunc(treeVN, &funcApp) && (funcApp.m_func == (VNFunc)GT_ADD))
+    {
+        if (vnStore->IsVNObjHandle(funcApp.m_args[0]) && vnStore->IsVNConstant(funcApp.m_args[1]))
```

Comment:
No hits for offset+handle so I didn't bother -- I was think of introducing a generic helper for this since I already needed it several times to accumulate the final offset

Reply:
That would be great. There are many places with this pattern (e. g. assertion prop).